# James O. Thomas

## Chapter 4 Homework

## Computational Physics

## Exercise 4.1

Write a program to calculate and print the factorial of a number entered by the user. If you wish you can base your program on the user-defined function for factorial given in section 2.6, but write your program so that it calculates the factorial using *integer* variables, not floating-point ones. use your program to calculate the factorial of 200.

Now modify your program to use floating-point variables instead and again calculate the factorial of 200. What do you find? Explain.

In [8]:
# Solution to 4.1

'''
Program to calculate the factorial of a number (n!)

Inputs: 
    n, the number to take the factorial of
Outputs:
    f, n!
    
James O. Thomas
September 2015
'''

import numpy as np
import math as math

def factorial(n):
    fact = 1
    for i in range(1,n+1):        # why does this not work when I use np.arange(1,n+1, dtype=int)?
        fact *= i
    return fact

n = int(input("Please input the integer value to calculate the factorial of: "))

n_fact = factorial(n)

print(n,'! = ', n_fact)
# print(n, '! = ', math.factorial(n))

Please input the integer value to calculate the factorial of: 200
200 ! =  788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000


When using integer data types, the user defined function returns the value of 200! as expected. There was no overflow because python can handle arbitrarily large integer values.

In [2]:
# Solution to 4.1

'''
Program to calculate the factorial of a number (n!) using floats as data types

Inputs: 
    n, the number to take the factorial of
Outputs:
    f, n!
    
James O. Thomas
September 2015
'''

import numpy as np
import math as math

def factorial(n):
    fact = 1.0
    for i in np.arange(1,n+1,1):
        fact = fact*i
    return fact

n = float(input("Please input the integer value to calculate the factorial of: "))

n_fact = factorial(n)

print(n,'! = ', n_fact)
print(n, '! = ', math.factorial(n))

Please input the integer value to calculate the factorial of: 200
200.0 ! =  inf
200.0 ! =  788657867364790503552363213932185062295135977687173263294742533244359449963403342920304284011984623904177212138919638830257642790242637105061926624952829931113462857270763317237396988943922445621451664240254033291864131227428294853277524242407573903240321257405579568660226031904170324062351700858796178922222789623703897374720000000000000000000000000000000000000000000000000


/home/james/anaconda3/lib/python3.4/site-packages/IPython/kernel/__main__.py:21: RuntimeWarning: overflow encountered in double_scalars


When using a floating point variable type, 200! return 'inf' which means that 200! is larger than the maximum allowed number for floats. ie $200! > 10^{308}$. This is what I would expect because a quick google search (as well as math.factorial(200)) has $200! \approx 7.9 \times 10^{374}$.

## Exercise 4.3: Calculating derivatives

Syppose we have a function $f(x)$ and we want to calculate its derivative at a point $x$. We can do that with pencil and paper if we know the mathematical form of the function, or we can do it on the computer by making use of the definition of the derivative:

$$\frac{df}{dx} = \lim_{\delta \to 0} \frac{f(x+\delta ) - f(x)}{\delta}$$

On the computer we can't actually tke the limit as $\delta $ goes to zero, but we can get a reasonable approximation just my making $\delta $ small.

a) Write a program that defines a function $f(x)$ returning the value $x(x-1)$, then calculate the derivative of the function at the point $x=1$ using the formula above with $\delta = 10^{-2}$. Calculate the true value of the same derivative analytically and compare with the answer your program gives. The two will not agree perfectly. Why not?

b) repeat the calculation for $\delta = 10^{-4}, 10^{-6}, 10^{-8}, 10^{-10}, 10^{-12}, 10^{-14}$. You should see that the accuracy of the calculation initially gets better as $\delta $ gets smaller, but then it gets worse again. Why is this?

We will look at numerical derivates in more detail in Section 5.10, where we will study techniques for dealing with these issues and maximizing the accuracy of our calculations.

In [24]:
# Solution to Exercise 4.3 a)
'''
Program to calculate derivatives

James O. Thomas
September 2015
'''

import numpy as np

def f(x):
    return x*(x-1)

delta = 1e-2
x_val = 1

fprime = (f(x_val + delta) - f(x_val))/delta

print(fprime)

1.010000000000001


Running the above code returns a value of $f'(1) \approx 1.01$

The analytical derivative:

$$f(x) = x(x-1) = x^2 -x $$

$$f'(x) = 2x-1 $$

$$f'(1) = 2(1) - 1 = 1$$

The two do not agree perfectly because of roundoff error associated with the python computed value. We also have a rather large value for $\delta $.

In [27]:
# Solution to Exercise 4.3 b)
'''
Program to calculate derivatives

James O. Thomas
September 2015
'''

import numpy as np

def f(x):
    return x*(x-1)

delta = np.array([1e-2, 1e-4, 1e-6, 1e-8, 1e-10, 1e-12])
x_val = 1

fprime = (f(x_val + delta) - f(x_val))/delta
error = (fprime - 1)

print('the calculated derivatives')
print(fprime)
print()
print('the errors')
print(error)

the calculated derivatives
[ 1.01        1.0001      1.000001    1.          1.00000008  1.0000889 ]

the errors
[  1.00000000e-02   9.99999999e-05   9.99917733e-07   3.92252875e-09
   8.28403710e-08   8.89005833e-05]


The errors initially get better because we are using a smaller and smaller $\delta $ value. This means that we are mathematically approaching the limit $\delta \to 0$. However, the errors get worse eventually because of roundoff errors associated with the $\delta $ value. Since the $ \delta $ value appears in the denominator of the definition of the derivative and is approaching 0, any small fluctuations in $\delta $ will have rather large effects on the derivative.

## Exercise 4.4: Calculating integrals

Suppose we want to calculate the value of the inegral

$$I = \int^1_{-1} \sqrt{1-x^2}dx $$

The integrand looks like a semicircle of radius 1: and hence the value of the inegral-the area under the curve-must be equal to $\frac{1}{2} \pi \approx 1.57079632679 ...$

Alternatively, we can evaluate the integral on the computer by dividing the domain of integration into a large number $N$ slices of width $h = 2/N$ each and then using the Riemann definition of the integral:

$$I = \lim_{N \to \infty } \sum^N_{k=1} hy_k $$

where

$$y_k = \sqrt{1-x_k^2} \rm{ and } x_k = -1 + hk $$

We cannot in practice take the limit $N \to \infty $, but we can make a reasonable approximation by just making $N$ large.

a) Write a program to evaluate the integral above with $ N = 100 $ and compare the result with the exact value. The two will not agree very well, because $N=100$ is not a sufficiently large number of slices.

b) increase the value of $N$ to get a more accurate value for the integral. if we require that the program runs in about one second or less, how accurate a value can you get?

Evaluating integrals is a common task in compuational physics calculations. We will study techniques for doing integrals in detail in the next chapter. As we will see, there are substantially quicker and more accurate methods than the simple one we have used here.

In [6]:
# solution to exercise 4.4
'''
Program to calculate the integral of a semicircle of radius 1

James O. Thomas
September 2015
'''

import numpy as np
import matplotlib.pyplot as plt

steps = 1e6
width = 2.0/steps

x = np.linspace(-1,1,steps)  # linspace includes both endpoints 
y = np.sqrt(1-x**2)

'''
plt.figure()
plt.plot(x,y)
plt.show()
'''

I = 0.0
for i in y:
    I = I + i*width
    
print('the calculated value is: ',I)
print('the absolute error is: ', I - 0.5*np.pi)
print('the percent error is: ', abs((I-0.5*np.pi)/(1.5*np.pi)*100))

the calculated value is:  1.57079475434
the absolute error is:  -1.57245935406e-06
the percent error is:  3.3368623868e-05


Part a) I get a percent error of around 0.37% when I use $N=100$.

Part b) For the code to run in about 1 second, I can use $10^{6}$ steps on my machine. With this number of steps I get an absolute error of around $10^{-6}$, and a percent error of about $3.3 \times 10^{-5}$. If I wanted to use this as part of a larger work, I would probably want to decrease the percent error down to about 1 part in a million or so, maybe more.